src/cross_sections.jl

```diff
@@ -0,0 +1,372 @@
+#
+# scattering probability
+#
+############
+
+function _unsafe_probability(
+    proc::AbstractProcessDefinition,
+    model::AbstractModelDefinition,
+    in_phase_space_def::AbstactPhasespaceDefinition,
+    in_phase_space::AbstractVector{T},
+    out_phase_space_def::AbstactPhasespaceDefinition,
+    out_phase_space::AbstractVector{T},
+)::Float64 where {T<:QEDbase.AbstractFourMomentum}
+    
+
+    matrix_elements_sq =
+        _matrix_element_square(proc, model, in_phase_space, out_phase_space)
+
+    normalization = averaging_norm(proc)
+
+    ps_fac = _phase_space_factor(proc,model,in_phasespace_def,in_phase_space,out_phasespace_def,out_phase_space)
+
+    return normalization *
+           sum(matrix_elements_sq) *
+            ps_fac
+end
+
+function _unsafe_probability(
+    proc::AbstractProcessDefinition,
+    model::AbstractModelDefinition,
+    in_phase_space_def::AbstactPhasespaceDefinition,
+    in_phase_space::AbstractVector{T},
+    out_phase_space_def::AbstactPhasespaceDefinition,
+    out_phase_space::AbstractMatrix{T},
+)::Float64 where {T<:QEDbase.AbstractFourMomentum}
+    res = Vector{eltype(T)}(undef,size(out_phase_space,2))
+    for i in 1:size(out_phase_space,2)
+        res[i] = _unsafe_probability(proc,model,in_phase_space_def,in_phase_space,out_phase_space_def,view(out_phase_space,:,i))
+    end
+    return res
+end
+
+function _unsafe_probability(
+    proc::AbstractProcessDefinition,
+    model::AbstractModelDefinition,
+    in_phase_space_def::AbstactPhasespaceDefinition,
+    in_phase_space::AbstractMatrix{T},
+    out_phase_space_def::AbstactPhasespaceDefinition,
+    out_phase_space::AbstractVector{T},
+)::Float64 where {T<:QEDbase.AbstractFourMomentum}
+    
+    res = Vector{eltype(T)}(undef,size(in_phase_space,2))
+    for i in 1:size(in_phase_space,2)
+        res[i] = _unsafe_probability(proc,model,in_phase_space_def,view(in_phase_space,:,i),out_phase_space_def,out_phase_space)
+    end
+    return res
+end
+
+function _unsafe_probability(
+    proc::AbstractProcessDefinition,
+    model::AbstractModelDefinition,
+    in_phase_space_def::AbstactPhasespaceDefinition,
+    in_phase_space::AbstractMatrix{T},
+    out_phase_space_def::AbstactPhasespaceDefinition,
+    out_phase_space::AbstractMatrix{T},
+)::Float64 where {T<:QEDbase.AbstractFourMomentum}
+    
+    res = Matrix{eltype(T)}(undef,size(in_phase_space,2), size(out_phase_space,2))
+    for i in 1:size(in_phase_space,2)
+        for j in 1:size(out_phase_space,2)
+            res[i,j] = _unsafe_probability(proc,model,in_phase_space_def,view(in_phase_space,:,i),out_phase_space_def,view(out_phase_space,:,j))
+        end
+    end
+    return res
+end
+
+function unsafe_probability(
+    proc::AbstractProcessDefinition,
+    model::AbstractModelDefinition,
+    in_phase_space_def::AbstactPhasespaceDefinition,
+    in_phase_space::AbstractVector{T},
+    out_phase_space_def::AbstactPhasespaceDefinition,
+    out_phase_space::AbstractVector{T},
```

Comment:
I think these should be `AbstractVecOrMat{T}` as well
```suggestion
    in_phase_space_def::AbstactPhasespaceDefinition,
    in_phase_space::AbstractVecOrMat{T},
    out_phase_space_def::AbstactPhasespaceDefinition,
    out_phase_space::AbstractVecOrMat{T},
```